mplc/multi_partner_learning.py

```diff
@@ -663,6 +664,577 @@ def fit_minibatch(self):
+                y_minibatch = tf.unstack(minibatch[1], axis=0)  # same here, with labels
+
+                for model_w, partner_w in zip(model.trainable_weights,
+                                              partners_weights[p_id]):  # set the model weights to partner's ones
+                    model_w.assign(partner_w.read_value())
+
+                for x, y in zip(x_minibatch, y_minibatch):  # iterate over batches
+                    with tf.GradientTape() as tape:
+                        y_pred = model(x)
+                        loss = model.compiled_loss(y, y_pred)
+                    model.compiled_metrics.update_state(y, y_pred)  # log the loss and accuracy
+                    partners_optimizers[p_id].minimize(loss, model.trainable_weights,
+                                                       tape=tape)  # perform local optimization
+
+                for model_w, partner_w in zip(model.trainable_weights,
+                                              partners_weights[p_id]):  # update the partner's weights
+                    partner_w.assign(model_w.read_value())
+            # at the end of the minibatch, aggregate all the local weights
+            aggregated_weights = aggregate_weights(tuple(partners_weights),
+                                                   agg_w)
+
+            # inform the partners with the new weights
+            for p_id in range(len(partners_minibatches)):
+                for new_w, partner_w in zip(aggregated_weights, partners_weights[p_id]):
+                    partner_w.assign(new_w.read_value())
+
+        # Execution
+        self.timer = time.time()
+        for e in range(self.epoch_count):
+            self.epoch_timer = time.time()
+            for partners_minibatches in zip(
+                    *self.train_dataset):  # <- partners_minibatches == [(x, y)] * nb_partners
+                fit_minibatch(self.model, partners_minibatches, self.partners_optimizers,
+                              self.partners_weights, self.agg_w)
+            epoch_history = self.get_epoch_history()  # compute val and train acc and loss.
+            # add the epoch history to self history, and log epoch number, and metrics values.
+            self.log_epoch(e, epoch_history)
+            self.epochs_index += 1
+            if self.early_stop():
+                break
+
+        self.log_end_training()
+
+    def early_stop(self):
+        """
+        If the metric monitored decreases (or increases, depends on the metrics monitored)
+        during patience epoch, we stop the learning.
+        """
+        if self.monitor_op(self.history[self.monitor_value][-1] - self.min_delta, self.best):
+            self.wait = 0
+            self.best = self.history[self.monitor_value][-1]
+        else:
+            self.wait += 1
+            if self.wait >= self.patience:
+                return True
+        return False
+
+
+class FastFedGrad(FastFedAvg):
+
```

Comment:
Can you please add a description `""" ... """` for the high-level understanding of what this approach does?

Reply:
It does exactly the same than FedGrad, but faster. I added a header to this part of the file, which describes that, but you're right I will add a docstring